appserver/payara-appserver-modules/microprofile/opentracing/src/main/java/fish/payara/microprofile/opentracing/jaxrs/JaxrsContainerRequestTracingFilter.java

```diff
@@ -235,26 +241,111 @@ public void filter(ContainerRequestContext requestContext, ContainerResponseCont
      * @return The name to use as the Span's operation name
      */
     private String determineOperationName(ContainerRequestContext requestContext, Traced tracedAnnotation) {
-        // If there is no annotation, or if there is an annotation and a config override providing an empty name
-        if (tracedAnnotation == null || ((String) OpenTracingCdiUtils.getConfigOverrideValue(
-                Traced.class, "operationName", resourceInfo, String.class)
-                .orElse(tracedAnnotation.operationName()))
-                .equals("")) {
-            // Set the name equal to the HTTP Method, followed by the method signature
-            return requestContext.getMethod()
-                    + ":"
-                    + resourceInfo.getResourceClass().getCanonicalName()
-                    + "."
+        // If there is no @Traced annotation
+        if (tracedAnnotation == null) {
+            Config config = null;
+            try {
+                config = ConfigProvider.getConfig();
+            } catch (IllegalArgumentException ex) {
+                logger.log(Level.INFO, "No config could be found", ex);
+            }
+
+            // Determine if an operation name provider has been given
+            Optional<String> operationNameProviderOptional = config.getOptionalValue(
+                    "mp.opentracing.server.operation-name-provider", String.class);
+            if (operationNameProviderOptional.isPresent()) {
+                String operationNameProvider = operationNameProviderOptional.get();
+
+                Path classLevelAnnotation = resourceInfo.getResourceClass().getAnnotation(Path.class);
+                Path methodLevelAnnotation = resourceInfo.getResourceMethod().getAnnotation(Path.class);
+
+                // If the provider is set to "http-path" and the @Path annotations are actually present
+                if (operationNameProvider.equals("http-path")
+                        && (classLevelAnnotation != null && methodLevelAnnotation != null)) {
+                    String operationName = requestContext.getMethod() + ":";
+
+                    if (classLevelAnnotation.value().startsWith("/")) {
+                        operationName += classLevelAnnotation.value();
+                    } else {
+                        operationName += "/" + classLevelAnnotation.value();
+                    }
+
+                    if (methodLevelAnnotation.value().startsWith("/")) {
+                        operationName += methodLevelAnnotation.value();
+                    } else {
+                        operationName += "/" + methodLevelAnnotation.value();
+                    }
+
+                    return operationName;
```

Comment:
What's the behaviour here if a class level annotation exists but a method level one doesn't?

Reply:
At this specific point? You wouldn't get in to the if block:
Line 264: `classLevelAnnotation != null && methodLevelAnnotation != null`

Reply:
Yeah I know it won't get into that block, I should really have commented a few lines down! I just mean does this method in general handle that scenario as expected? Because I don't see any TCK tests for that in particular so it may not come up!

Reply:
Atm it requires that both be set, otherwise it reverts to using the class-method style.
And no the TCK doesn't check for this behaviour and the spec doesn't state what should happen.

I'll check if just using the Path annotation from the class breaks the TCK.